### PR TITLE
feat: support user-defined custom check runs 

### PR DESCRIPTION
**Pull request by rnetser**
The feature allows users to define custom checks in their YAML config that:
  - Execute uv tool run commands securely
  - Support configurable triggers (opened, synchronize, reopened, etc.)
  - Include secret redaction in logs
  - Integrate with GitHub Check Runs API

## Summary by CodeRabbit

* **New Features**
  * Per-repo configurable custom check runs shown in welcome/retest lists, runnable via /retest, and reported as PR status checks; built-in checks protected.

* **Refactor**
  * Unified, name-based check-run lifecycle and standardized status payloads; custom checks integrated into CI/retest flows.

* **Tests**
  * Extensive coverage for schema, validation, execution, env handling, retests, and edge cases.

* **Documentation**
  * Added Security guidance for custom check runs in README.

* **Style**
  * Exposed duration-formatting helper for build summaries.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: webhook_server/libs/handlers/runner_handler.py [hunk 1]
```diff
@@ -569,3 +644,41 @@ async def cherry_pick(self, pull_request: PullRequest, target_branch: str, revie
...
+        tasks: list[Coroutine[Any, Any, Any] | Task[Any]] = []
+        for _test in supported_retests:
+            self.logger.debug(f"{self.log_prefix} running retest {_test}")
+            task = asyncio.create_task(_retests_to_func_map[_test](pull_request=pull_request))
+            tasks.append(task)
+
+        results = await asyncio.gather(*tasks, return_exceptions=True)
+        for result in results:
+            if isinstance(result, Exception):
+                self.logger.error(f"{self.log_prefix} Async task failed: {result}")
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/libs/handlers/runner_handler.py]: Done

File: webhook_server/tests/test_github_api.py [hunk 2]
```diff
@@ -1415,7 +1419,13 @@ async def test_clone_repository_empty_checkout_ref(
...
             mock_config.repository_data = {"enabled": True}
-            mock_config.get_value.return_value = None
+
+            def get_value_side_effect(value: str, *_args: Any, **_kwargs: Any) -> Any:
+                if value == "custom-check-runs":
+                    return []
+                return None
+
+            mock_config.get_value.side_effect = get_value_side_effect
             mock_config.data_dir = str(tmp_path)
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/tests/test_github_api.py]: Done

File: webhook_server/config/schema.yaml [hunk 3]
```diff
@@ -318,3 +318,53 @@ properties:
...
+                description: |
+                  Environment variables to pass to the command.
+                  Each entry must be in 'VAR_NAME=value' format.
+                items:
+                  type: string
+                  pattern: "^[^=]+=.+$"
+            required:
+              - name
+              - command
+            additionalProperties: false
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/config/schema.yaml]: Done

File: webhook_server/libs/github_api.py [hunk 4]
```diff
@@ -622,6 +622,12 @@ def _repo_data_from_config(self, repository_config: dict[str, Any]) -> None:
             value="pre-commit", return_on_none=False, extra_dict=repository_config
         )
 
+        # Load and validate custom check runs
+        raw_custom_checks = self.config.get_value(
+            value="custom-check-runs", return_on_none=[], extra_dict=repository_config
+        )
+        self.custom_check_runs: list[dict[str, Any]] = self._validate_custom_check_runs(raw_custom_checks)
+
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/libs/github_api.py]: Done

File: webhook_server/tests/test_github_api.py [hunk 5]
```diff
@@ -1640,3 +1650,63 @@ def rmtree_fail(*args, **kwargs):
...
+                                    "'build-container' conflicts with built-in check" in call for call in warning_calls
+                                )
+                                assert any(
+                                    "'python-module-install' conflicts with built-in check" in call
+                                    for call in warning_calls
+                                )
+                                assert any(
+                                    "'conventional-title' conflicts with built-in check" in call
+                                    for call in warning_calls
+                                )
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/tests/test_github_api.py]: Done. Changed varargs typing from `Any` to `object` for unused parameters. Removed unused `logger` fixture parameter from test functions.

**Comment by rnetser** [comment]
@coderabbitai

pushing another fix now; review it thoroughly and not in chunks

File: webhook_server/tests/test_github_api.py [hunk 6]
```diff
@@ -1640,3 +1650,104 @@ def rmtree_fail(*args, **kwargs):
...
+
```
Comment (rnetser) [inline webhook_server/tests/test_github_api.py]: Done

File: webhook_server/tests/test_github_api.py [hunk 7]
```diff
@@ -1640,3 +1650,104 @@ def rmtree_fail(*args, **kwargs):
...
+                                # Verify that only unique names are kept (first occurrence wins)
+                                assert len(gh.custom_check_runs) == 2
+                                check_names = [c["name"] for c in gh.custom_check_runs]
+                                assert "my-check" in check_names
+                                assert "another-check" in check_names
+
+                                # Verify that warning was logged for duplicate
+                                warning_calls = [str(call) for call in mock_logger.warning.call_args_list]
+                                assert any("Duplicate custom check name 'my-check'" in call for call in warning_calls)
```
Comment (rnetser) [inline webhook_server/tests/test_github_api.py]: Done

File: webhook_server/tests/test_custom_check_runs.py [hunk 8]
```diff
@@ -0,0 +1,860 @@
...
+                new=AsyncMock(return_value=(True, "Lint passed", "")),
+            ),
+        ):
+            # Execute the check
+            await runner_handler.run_custom_check(pull_request=mock_pull_request, check_config=check_config)
+
+            # Verify workflow: in_progress -> execute -> success
+            runner_handler.check_run_handler.set_custom_check_in_progress.assert_called_once()
+            runner_handler.check_run_handler.set_custom_check_success.assert_called_once()
+
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/tests/test_custom_check_runs.py]: code was changed; closing

File: webhook_server/tests/test_github_api.py [hunk 9]
```diff
@@ -1415,7 +1419,13 @@ async def test_clone_repository_empty_checkout_ref(
...
             mock_config = Mock()
             mock_config.repository_data = {"enabled": True}
-            mock_config.get_value.return_value = None
+
+            def get_value_side_effect(value: str, *_args: object, **_kwargs: object) -> list[object] | None:
+                if value == "custom-check-runs":
+                    return []
+                return None
+
+            mock_config.get_value.side_effect = get_value_side_effect
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/tests/test_github_api.py]: Done

**Comment by rnetser** [comment]
@coderabbitai
please review this pr throughly and if you have nay, add all commnents at once

File: webhook_server/config/schema.yaml [hunk 10]
```diff
@@ -318,3 +318,53 @@ properties:
...
+
+            Environment Variables:
+            The 'env' field supports explicit variable assignment using 'VAR_NAME=value' format.
+            Each entry must include both the variable name and value separated by '='.
+
+            Examples:
+              - name: lint
+                command: uv tool run --from ruff ruff check
+                env:
+                  - TOKEN=xyz
```
Comment (myakove) [inline webhook_server/config/schema.yaml]: I don't think TOKEN is relevant for this example.
Also do not use env, since commands can run in parallel in overwrite each other envs

File: webhook_server/libs/github_api.py [hunk 11]
```diff
@@ -754,6 +760,14 @@ def _current_pull_request_supported_retest(self) -> list[str]:
 
         if self.conventional_title:
             current_pull_request_supported_retest.append(CONVENTIONAL_TITLE_STR)
+
+        # Add custom check runs
+        # Note: custom checks are validated in _validate_custom_check_runs()
+        # so name is guaranteed to exist
+        for custom_check in self.custom_check_runs:
+            check_name = custom_check["name"]
+            current_pull_request_supported_retest.append(check_name)
```
Comment (myakove) [inline webhook_server/libs/github_api.py]: Do we see it in the welcome msg? if not we should see it.

File: webhook_server/libs/handlers/check_run_handler.py [hunk 12]
```diff
@@ -213,6 +213,22 @@ async def set_cherry_pick_failure(self, output: dict[str, Any]) -> None:
...
+
+    async def set_custom_check_success(self, name: str, output: dict[str, str] | None = None) -> None:
+        """Set custom check run to success."""
+        await self.set_check_run_status(check_run=name, conclusion=SUCCESS_STR, output=output)
+
+    async def set_custom_check_failure(self, name: str, output: dict[str, str] | None = None) -> None:
+        """Set custom check run to failure."""
+        await self.set_check_run_status(check_run=name, conclusion=FAILURE_STR, output=output)
+
     async def set_check_run_status(
```
Comment (myakove) [inline webhook_server/libs/handlers/check_run_handler.py]: This can be used now for all the existing check runs, right? You can remove all others and use only those.

File: webhook_server/libs/handlers/pull_request_handler.py [hunk 13]
```diff
@@ -619,6 +619,14 @@ async def process_opened_or_synchronize_pull_request(self, pull_request: PullReq
         if self.github_webhook.conventional_title:
             setup_tasks.append(self.check_run_handler.set_conventional_title_queued())
 
+        # Queue custom check runs (same as built-in checks)
+        # Note: custom checks are validated in GithubWebhook._validate_custom_check_runs()
+        # so name is guaranteed to exist
+        for custom_check in self.github_webhook.custom_check_runs:
```
Comment (myakove) [inline webhook_server/libs/handlers/pull_request_handler.py]: After you add the custom commands it should be same as the others check runs
the flow should be:
you read the config, validate the commands, add them to the existing checkruns and from there it should be the same flow as the built in checkruns

**Review by myakove — CHANGES_REQUESTED** [review]

File: webhook_server/libs/github_api.py (hunk 11 above)
Comment (rnetser) [reply webhook_server/libs/github_api.py]: will add

File: webhook_server/config/schema.yaml [hunk 14]
```diff
@@ -318,3 +318,41 @@ properties:
...
+            Examples:
+              - name: lint
+                command: uv tool run --from ruff ruff check
+                mandatory: true
+              - name: security-scan
+                command: TOKEN=xyz DEBUG=true uv tool run --from bandit bandit -r .
+                mandatory: false
+              - name: complex-check
+                command: |
+                  python -c "
```
Comment (myakove) [inline webhook_server/config/schema.yaml]: `uv run python -c`

File: webhook_server/libs/handlers/runner_handler.py [hunk 15]
```diff
@@ -435,13 +431,13 @@ async def run_conventional_title_check(self, pull_request: PullRequest) -> None:
...
 
-        await self.check_run_handler.set_conventional_title_in_progress()
+        await self.check_run_handler.set_check_in_progress(name=CONVENTIONAL_TITLE_STR)
         allowed_names = [name.strip() for name in self.github_webhook.conventional_title.split(",") if name.strip()]
         title = pull_request.title
 
         self.logger.debug(f"{self.log_prefix} Conventional title check for title: {title}, allowed: {allowed_names}")
         if any([re.match(rf"^{re.escape(_name)}(\([^)]+\))?!?: .+", title) for _name in allowed_names]):
-            await self.check_run_handler.set_conventional_title_success(output=output)
+            await self.check_run_handler.set_check_success(name=CONVENTIONAL_TITLE_STR, output=output)
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/libs/handlers/runner_handler.py]: Done

File: webhook_server/libs/handlers/runner_handler.py [hunk 16]
```diff
@@ -565,7 +587,49 @@ async def cherry_pick(self, pull_request: PullRequest, target_branch: str, revie
...
+                self.logger.error(f"{self.log_prefix} Unknown retest '{_test}' requested, skipping")
+                continue
+            self.logger.debug(f"{self.log_prefix} running retest {_test}")
+            task = asyncio.create_task(runner(pull_request=pull_request))
+            tasks.append(task)
+
+        results = await asyncio.gather(*tasks, return_exceptions=True)
+        for result in results:
+            if isinstance(result, Exception):
+                self.logger.error(f"{self.log_prefix} Async task failed: {result}")
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/libs/handlers/runner_handler.py]: Done

File: webhook_server/tests/test_custom_check_runs.py [hunk 17]
```diff
@@ -0,0 +1,912 @@
...
+        # Mock shutil.which to find the full path executable
+        with patch("shutil.which", return_value="/usr/local/bin/custom_tool"):
+            validated = GithubWebhook._validate_custom_check_runs(mock_github_webhook, raw_checks)
+
+        # Should validate successfully
+        assert len(validated) == 1
+        assert validated[0]["name"] == "full-path"
+
+
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/tests/test_custom_check_runs.py]: Done

File: webhook_server/libs/github_api.py [hunk 18]
```diff
@@ -860,6 +875,123 @@ async def cleanup(self) -> None:
...
+        - Verifies command executable exists on server using shutil.which()
+        - Logs warnings for invalid checks and skips them
+
+        Args:
+            raw_checks: List of custom check configurations from config
+
+        Returns:
+            List of validated custom check configurations
+        """
+        validated_checks: list[dict[str, Any]] = []
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/libs/github_api.py]: Done

File: webhook_server/libs/github_api.py [hunk 19]
```diff
@@ -860,6 +875,123 @@ async def cleanup(self) -> None:
...
+
+        # Regex to match env-var assignments (e.g., FOO=bar, MY_VAR=123)
+        env_assign_re = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*=")
+
+        for check in raw_checks:
+            # Validate name field
+            check_name = check.get("name")
+            if not check_name:
+                self.logger.warning("Custom check missing required 'name' field, skipping")
+                continue
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/libs/github_api.py]: Done

File: webhook_server/utils/command_security.py [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+    (r"/boot/", "Access to /boot/ is not allowed"),
+    (r"\.\.\/", "Path traversal (..) is not allowed"),
+    (r"\.env", "Access to .env files is not allowed"),
+    (r"config\.yaml", "Access to config.yaml is not allowed"),
+    (r"credentials", "Access to credentials files is not allowed"),
+    (r"\.pem\b", "Access to PEM files is not allowed"),
+    (r"\.key\b", "Access to key files is not allowed"),
+    (r"id_rsa", "Access to SSH private keys is not allowed"),
+    (r"id_ed25519", "Access to SSH private keys is not allowed"),
+]
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/utils/command_security.py]: Acknowledged - these nitpicks were addressed in subsequent commits.

File: webhook_server/tests/test_custom_check_runs.py [hunk 21]
```diff
@@ -0,0 +1,936 @@
...
+            env_dict = call_kwargs["env"]
+            assert env_dict is not None
+            assert env_dict["DEBUG"] == "true"
+            assert env_dict["VERBOSE"] == "1"
+            # Verify parent environment is inherited
+            assert "PATH" in env_dict
+            # INVALID_VAR should not be in env_dict (skipped)
+            assert "INVALID_VAR" not in env_dict
+            # INVALID_VAR should be skipped and a warning logged
+            runner_handler.logger.warning.assert_called()
```
[the remark this answers is not in the document]
Comment (rnetser) [reply webhook_server/tests/test_custom_check_runs.py]: Acknowledged - test refactored to use CUSTOM_CHECK_PREFIX constant and tmp_path fixture.

File: webhook_server/libs/handlers/check_run_handler.py [hunk 22]
```diff
@@ -427,7 +375,16 @@ async def all_required_status_checks(self, pull_request: PullRequest) -> list[st
         if self.github_webhook.conventional_title:
             all_required_status_checks.append(CONVENTIONAL_TITLE_STR)
 
-        _all_required_status_checks = branch_required_status_checks + all_required_status_checks
+        # Add mandatory custom checks only (default is mandatory=true for backward compatibility)
```
Comment (myakove) [inline webhook_server/libs/handlers/check_run_handler.py]: `backward compatibility)` your PR introduce this so what is backward compatibility?

**Comment by rnetser** [comment]
/verified

- retest command
- welcome message
- check is added as optional to PRs

**Comment by myakove** [comment]
rebase so we can merge it @rnetser
